src/coprocessor/statistics/histogram.rs

```diff
@@ -0,0 +1,237 @@
+        bucket.set_repeats(self.repeats as i64);
+        bucket.set_count(self.count as i64);
+        bucket.set_lower_bound(self.lower_bound);
+        bucket.set_upper_bound(self.upper_bound);
+        bucket
+    }
+}
+
+/// Histogram represents statistics for a column or index.
+#[derive(Default)]
+pub struct Histogram {
+    // number of distinct values
+    ndv: u64,
+    buckets: Vec<Bucket>,
+    // max number of values in per bucket. Notice: when a bucket's count is equal to
+    // per_bucket_limit, only value equal with last value can been inserted into it.
+    per_bucket_limit: u64,
+    // max number of buckets
+    buckets_num: usize,
+}
+
+impl Histogram {
+    pub fn new(buckets_num: usize) -> Histogram {
+        Histogram {
+            per_bucket_limit: 1,
+            buckets_num: buckets_num,
+            ..Default::default()
+        }
+    }
+
+    pub fn into_proto(self) -> analyze::Histogram {
+        let mut hist = analyze::Histogram::new();
+        hist.set_ndv(self.ndv as i64);
+        let buckets: Vec<analyze::Bucket> = self.buckets
+            .into_iter()
+            .map(|bucket| bucket.into_proto())
+            .collect();
+        hist.set_buckets(RepeatedField::from_vec(buckets));
+        hist
+    }
+
+    // insert a data bigger than or equal to the max value in current histogram.
+    pub fn append(&mut self, data: Vec<u8>) {
+        if let Some(bucket) = self.buckets.last_mut() {
+            // The new item has the same value as last bucket value, to ensure that
+            // a same value only stored in a single bucket, we do not increase bucket
+            // even if it exceeds per_bucket_limit.
+            if bucket.upper_bound == data {
+                bucket.append_repeated_item();
+                return;
+            }
+        }
+        self.ndv += 1;
+        if self.buckets.len() >= self.buckets_num && self.is_last_bucket_full() {
+            self.merge_buckets();
+        }
+
+        if !self.is_last_bucket_full() {
+            self.buckets.last_mut().unwrap().append_diff_item(data);
+            return;
+        }
+
+        // create a new bucket and insert data
+        let mut count = 1;
+        if let Some(bucket) = self.buckets.last() {
+            count += bucket.count
+        }
+        self.buckets.push(Bucket::new(count, data.clone(), data, 1));
+    }
+
+    // check whether the last bucket is full.
+    fn is_last_bucket_full(&self) -> bool {
+        if self.buckets.is_empty() {
+            return true;
+        }
+        let count = if self.buckets.len() == 1 {
+            self.buckets[0].count
+        } else {
+            let len = self.buckets.len();
+            self.buckets[len - 1].count - self.buckets[len - 2].count
+        };
+        count >= self.per_bucket_limit
+    }
+
+    // It merges every two neighbor buckets.
+    fn merge_buckets(&mut self) {
+        let bucket_num = (self.buckets_num >> 1) + (self.buckets_num & 1);
```

Comment:
It's 21st century, compiler knows how to optimize it, please write readable code.